Adding extended support for `Box`, `Rc`, and `Arc`

LucaCappelletti94 opened the pull request:
This PR addresses discussion #4609

Still have to: 

* [x] Implement also:
    * [x] `impl<T, DB, const N: usize> CanInsertInSingleQuery<DB> for Rc<[T; N]>`
    * [x] `impl<T, DB, const N: usize> CanInsertInSingleQuery<DB> for Arc<[T; N]>`
    * [x] `impl<DB: Backend> Migration<DB> for Rc<dyn Migration<DB> + '_> `
    * [x] `impl<DB: Backend> Migration<DB> for Arc<dyn Migration<DB> + '_> `
    * [x] `impl<T: QueryId + ?Sized> QueryId for Rc<T>`
    * [x] `impl<T: QueryId + ?Sized> QueryId for Arc<T>`
    * [x] `impl<T: ?Sized, DB> QueryFragment<DB> for Rc<T>`
    * [x] `impl<T: ?Sized, DB> QueryFragment<DB> for Arc<T>`
    * [x] `impl<T: Insertable> Insertable for Rc<T>`
    * [x] `impl<T: Insertable> Insertable for Arc<T>`
    * [x] `impl<T: AppearsOnTable> AppearsOnTable for Rc<T>`
    * [x] `impl<T: AppearsOnTable> AppearsOnTable for Arc<T>`
    * [x] `impl<T: SelectableExpression> SelectableExpression for Rc<T>`
    * [x] `impl<T: SelectableExpression> SelectableExpression for Arc<T>`
    * [x] `impl<T: ValidGrouping> ValidGrouping for Rc<T>`
    * [x] `impl<T: ValidGrouping> ValidGrouping for Arc<T>`
    * [x] `impl<T: ?Sized, ST, DB> Queryable<ST, DB> for std::rc::Rc<T>`
    * [x] `impl<T: ?Sized, ST, DB> Queryable<ST, DB> for std::sync::Arc<T>`
    * [x] `impl<T: ?Sized, ST, DB> Queryable<ST, DB> for Box<T>`
    * [x] `impl<impl<T: ?Sized, ST, DB> ToSql<ST, DB> for std::rc::Rc<T>`
    * [x] `impl<impl<T: ?Sized, ST, DB> ToSql<ST, DB> for std::sync::Arc<T>`
    * [x] `impl<impl<T: ?Sized, ST, DB> ToSql<ST, DB> for Box<T>`
    * [x] `impl<T, ST, DB> FromSql<ST, DB> for std::rc::Rc<T>`
    * [x] `impl<T, ST, DB> FromSql<ST, DB> for std::sync::Arc<T>` 
    * [x] `impl<T, ST, DB> FromSql<ST, DB> for Box<T>` 
* [x] Extend test suite for `Rc`
* [x] Extend test suite for `Arc`

weiznich commented:
Thanks for opening this PR. I think we likely want to cover most of [these](https://github.com/search?q=repo%3Adiesel-rs%2Fdiesel+%22for+Box%3C%22&type=code) impls as well. (All but that one for `Instrumentation`)

LucaCappelletti94 commented:
@weiznich I have extended the list of traits to be implemented, please do let me know whether anything is missing.

weiznich commented:
From the search linked above it's missing the following impls:

* `Insertable for Arc/Rc`
* `Migration` for Arc/Rc`
* `AppearsOnTable for Arc/Rc`
* `SelectableExpression for Arc/Rc`
* `ValidGrouping for Arc/Rc`

Also it lists the instrumentation impl which will not be helpful as the only method there takes a `&mut self` argument which you won't have for Arc/Rc for obvious reasons.

LucaCappelletti94 commented:
> From the search linked above it's missing the following impls:
> 
>     * `Insertable for Arc/Rc`
> 
>     * `Migration for Arc/Rc`
> 
>     * `AppearsOnTable for Arc/Rc`
> 
>     * `SelectableExpression for Arc/Rc`
> 
>     * `ValidGrouping for Arc/Rc`
> 
> 
> Also it lists the instrumentation impl which will not be helpful as the only method there takes a `&mut self` argument which you won't have for Arc/Rc for obvious reasons.

I have updated the PR tasks to include also these traits and removed the `Instrumentation`

LucaCappelletti94 commented:
I am afraid I will require some more instructions regarding how to proceed. As you can see from the [CI run](https://github.com/diesel-rs/diesel/actions/runs/15212136723/job/42788661444?pr=4610), trying to use `std::rc::Rc` or `std::sync::Arc` in a struct leads to the following errors:

```
 the trait `diesel::Expression` is not implemented for `std::string::String`
```

and

```
the trait `diesel::AppearsOnTable<schema::comments::table>` is not implemented for `std::string::String`
```

Which make sense as indeed, those traits are needed for the current trait implementations and those traits are not implemented for String. What should be preferable to do?

weiznich commented:
> Just to better understand, https://github.com/diesel-rs/diesel/discussions/4609 I was asking exactly regarding the use of Rc in such structs, but maybe I had explained myself poorly. Do you believe it should be viable to do, in theory, or are there any blocking reasons?

That was mostly me skipping over details assuming I know what you want there. 

That written, it's actually good to know that we cannot have both, which means from my point of view this is for now kind of a dead end as I consider both use-cases (using `Rc<dyn BoxableExpression>` and `Rc<String` as field in your struct) as something that would be desirable to support. If we cannot easily support both it's kind of hard to say which one should be supported without making it essentially impossible to support the other one as well. So as long as we do not have a solution for this I would hold of implementing anything to not block any future extensions. I personally do currently not have the time to investigate this in depth, given the many other things I need to handle (like answering all your questions).

> That being said, there is a [blanked implementation](https://github.com/diesel-rs/diesel/blob/db7edfe9db4313dd3f400dcdd0ab2180287ac2eb/diesel/src/expression/mod.rs#L232-L242) adding AsExpression<ST> to all types which already implement Expression<ST> for that ST, and Rc<T> would therefore implement AsExpression if it can implement Expression, which depends on T having implemented Expression - why does not T=String (in this case) implement the Expression trait?

Note that this only implements `AsExpression` if the underlying type is an actual `Expression` `String` or any other rust value type is never considered a SQL expression by diesel, so that wild card impl is not helpful here to support both variants. What might be viable is to generate specific `Rc` and `Arc` variants for the `AsExpression` impls via the derive, but again that would drastically increase the number of impls generated by diesel, which in turn will certainly have an impact on compile times and error messages. It might be worth to quantify that impact, maybe it's not as bad as assumed but without hard numbers I wouldn't like blindly  to go that way.

LucaCappelletti94 commented:
Thank you for your reply - regarding extending the `AsExpression` derive, I can try to add the `Rc` and `Arc` derive impls and report how much the compile time increase for both diesel and the practical case of my project.

I am not sure I understood whether I should try to add support for `impl<ST> AsExpression<ST> for Rc<String> where String: AsExpression<ST>` (and other std/core types), or whether such an implementation would preclude potential future impls of `Rc<dyn BoxableExpression>` and as such should not be added for the time being. I suppose you mentioned extending the `AsExpression` derive exactly because I could then wrap the `String` type into a custom type?

Thank you again for the patience in answering all of these questions, hopefully over time I will learn better the internals of `diesel` and be better able to contribute to it.

weiznich commented on diesel_derives/src/lib.rs:
```diff
@@ -139,6 +139,10 @@ pub fn derive_as_changeset(input: TokenStream) -> TokenStream {
 /// - `impl AsExpression<Nullable<SqlType>> for YourType`
 /// - `impl AsExpression<SqlType> for &'a YourType`
 /// - `impl AsExpression<Nullable<SqlType>> for &'a YourType`
+/// - `impl AsExpression<SqlType> for std::rc::Rc<YourType>`
+/// - `impl AsExpression<SqlType> for &'a std::rc::Rc<YourType>`
+/// - `impl AsExpression<SqlType> for std::sync::Arc<YourType>`
+/// - `impl AsExpression<SqlType> for &'a std::sync::Arc<YourType>`
```
That should be removed if it's not generated by this derive anymore

weiznich commented on diesel/src/type_impls/primitives.rs:
```diff
@@ -271,3 +271,119 @@ where
     }
 }
+
+impl<T: ?Sized, ST, DB> ToSql<ST, DB> for std::rc::Rc<T>
```
I would like to add `Box<T>` variants for these impls as well as `Box` is essentially just another smart pointer.

LucaCappelletti94 replied:
For all traits described in the first post of the PR, or only for specifically `ToSql`?

LucaCappelletti94 replied:
Yeah as expected, I can use `Rc` and `Arc` in the structs but the same does not work for `Box`. Again, it is not clear to me why the blanked covers the `Box` case but does not cover the `Arc` and `Rc`, I believe it has something to do with Rust's auto-dereferencing.

weiznich commented on diesel_tests/tests/schema/mod.rs:
```diff
@@ -90,6 +90,44 @@ impl Comment {
+    pub id: i32,
+    pub post_id: i32,
+    pub text: std::sync::Arc<String>,
+}
```
Could we have a simple test here that inserts + loads both variants?

weiznich commented on diesel_tests/tests/schema/mod.rs:
```diff
@@ -90,6 +90,44 @@ impl Comment {
+pub struct CommentRc {
+    pub id: i32,
+    pub post_id: i32,
+    pub text: std::rc::Rc<str>,
```
I would expect that this causes issues for loading values as we don't have a `ToSql<DB, Text> for str` impl yet. That said it seems reasonable to support this as well, but I would like to keep that off this PR to not expand it's scope any further. So just go with `Rc<String>` for now.

weiznich reviewed (commented):
Thanks for updating this PR.

As these impls do not seem to conflict with anything it should be fine to go that way.

There are still a few things I noticed about the changes. 

Additionally I would like to see the following things added to the PR:

- A changelog entry stating that `Rc`+`Arc` are now supported types for serialization + deserialization + `BoxableExpression`
- A testcase that actually verifies inserting + loading data works
- A test case that verifies that `Arc<dyn BoxableExpression>` (and the `Rc` equivalent works)

> That being said, it turns out that methods like get_result and first require Send and since Rc does not implement Send, I get the error  `Rc<str>` cannot be sent between threads safely. At least, these are required when using diesel-async, I am not sure whether the same would apply also for diesel in its sync version.

That's `diesel-async` specific. It should just work with normal diesel, so it's reasonable to keep it here.

> Do you think such Send trait requirements could be removed in any way?

Not without regressing use-cases that require the `Send` bound to be there as there is currently no way to abstract over the fact whether or not a type implements `Send`.

LucaCappelletti94 commented:
I am trying to test and fix the case for `Rc<dyn BoxableExpression>`, but unfortunately this case does not seem to implement Expression: `Rc<dyn BoxableExpression<table, Pg, SqlType = _>>`. Same thing happens for `Arc`.

```rust
#[diesel_test_helper::test]
fn can_rc_query_with_boxable_expression() {
    let connection = &mut connection_with_sean_and_tess_in_users_table();

    let expr: Rc<dyn BoxableExpression<users::table, _, SqlType = _>> =
        Rc::new(users::name.eq("Sean")) as _;

    let data = users::table.into_boxed().filter(expr).load(connection);
    let expected = vec![find_user_by_name("Sean", connection)];
    assert_eq!(Ok(expected), data);

    let expr: Rc<dyn BoxableExpression<users::table, _, SqlType = _>> =
        Rc::new(users::name.eq("Sean")) as _;

    let data = users::table.filter(expr).into_boxed().load(connection);
    let expected = vec![find_user_by_name("Sean", connection)];
    assert_eq!(Ok(expected), data);
}
```

with the full error being:

```rust
error[E0277]: the trait bound `Rc<dyn BoxableExpression<table, Pg, SqlType = _>>: AppearsOnTable<...>` is not satisfied
   --> diesel_tests/tests/boxed_queries.rs:201:42
    |
201 |     let data = users::table.into_boxed().filter(expr).load(connection);
    |                                          ^^^^^^ the trait `diesel::Expression` is not implemented for `Rc<dyn diesel::BoxableExpression<schema::users::table, Pg, SqlType = _>>`
    |
    = help: the following other types implement trait `diesel::Expression`:
              &T
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
              (T0, T1, T2, T3, T4, T5, T6)
              (T0, T1, T2, T3, T4, T5, T6, T7)
            and 452 others
    = note: required for `Rc<dyn diesel::BoxableExpression<schema::users::table, Pg, SqlType = _>>` to implement `diesel::AppearsOnTable<schema::users::table>`
    = note: required for `BoxedSelectStatement<'_, (Integer, Text, Nullable<Text>), ..., ...>` to implement `FilterDsl<Rc<dyn diesel::BoxableExpression<schema::users::table, Pg, SqlType = _>>>`
    = note: the full name for the type has been written to '/home/luca/github/diesel/target/debug/deps/integration_tests-14ee3c308f18e762.long-type-688333457155762760.txt'
    = note: consider using `--verbose` to print the full type name to the console
```

LucaCappelletti94 commented:
Ran the benchmark between this branch and current main.

This time around, I generated a synthetic but plausible schema with 50 tables, each with randomly 6-12 columns, and each with random types from the set of types supported by sqlite & postgres.

Each profile (`dev` and `release`) was built five times from a cold cache (`cargo clean` between runs) and the reported value is the median. Timing comes from `/usr/bin/time -f "%e %U %S %M"` (wall, user CPU, system CPU, peak RSS). The host is an AMD Threadripper PRO 5975WX (64 threads, 1 TB RAM).

| Profile | Branch | Wall | User | Max RSS |
| --- | --- | ---: | ---: | ---: |
| dev | `upstream/main` | 37.67 s | 42.82 s | 1 230 MiB |
| | `rc-arc` | 38.04 s | 43.38 s | 1 238 MiB |
| release | `upstream/main` | 38.70 s | 50.73 s | 1 358 MiB |
| | `rc-arc` | 38.32 s | 49.99 s | 1 361 MiB |

Wall-clock values across all 5 runs per cell (s, sorted), with transient system spike during one run on each branch marked with †. I am unsure what that is.

| Profile | `upstream/main` | `rc-arc` |
| --- | --- | --- |
| dev | 36.71 / 37.00 / **37.67** / 37.88 / 39.02 | 35.89 / 38.03 / **38.04** / 38.45 / 39.16 |
| release | 37.50 / 37.88 / **38.70** / 39.01 / 60.95† | 36.90 / 37.39 / **38.32** / 38.83 / 60.84† |

All four median deltas land between −1.5% and +1.3%, so I would say it does not cause a meaningful change in this setup.

weiznich reviewed (approved):
This looks good now, thanks for the update

Would it be possible to squash the commits + add a change log entry for this feature before merging.

I hopefully get around to review at least some of your other PR's in the next weeks as well if nothing important shows up again (like more security issues :()

LucaCappelletti94 committed:
Add support for `Box<T>`, `Rc<T>`, and `Arc<T>`

LucaCappelletti94 commented:
Done!
